### DISCUSSION TITLE
Support OAuth/OpenID Connect Login

### DISCUSSION THREAD

**Discussion by skuzzle** [jonaswinkler/paperless-ng#639] [Feature Requests]
Hi,

it would be great if paperless-ng supported being an OAuth client so I can connect it to my authorization server.

I tried to achieve OAuth login using [traefik-forward-auth](https://github.com/thomseddon/traefik-forward-auth) but that requires paperless to inspect the `X-Forwarded-User` header. This seems to be closely related to the already supported `HTTP_REMOTE_USER` option but seems to require some manual coding in order to inspect HTTP headers, as described in the [Django documentation](https://docs.djangoproject.com/en/3.1/howto/auth-remote-user/)

**Comment by jonaswinkler** [comment]
Isn't that really just what's already in paperless with `PAPERLESS_ENABLE_HTTP_REMOTE_USER`, but with a different header?

I don't know mutch about Oauth, so I really can't provide any input here. There's also this warning in the documentation to consider.

**Reply by skuzzle** [reply]
I guess when `PAPERLESS_ENABLE_HTTP_REMOTE_USER` is enabled the application expects the authenticated username in a _Environment variable_ called `REMOTE_USER`. Thus it has to somehow be set by the webserver or the the system that is executing the application. I'm not quite sure either how this is intended do work.
However, the Django documentation that I linked to shows a way to extract the `REMOTE_USER` from an actual HTTP header (and also mentions all the security implications this has)

**Reply by jonaswinkler** [reply]
Hm. With PAPERLESS_ENABLE_HTTP_REMOTE_USER enabled I'm able to authenticate by simply sending `Remote-User: anyusername`, and the API works. @shamoon you might be interested in this.

So this can be configured to accept `X-Forwarded-User` instead.

I believe that's also what the warning is for. Don't forward any `Remote-User` headers (or whatever paperless accepts) to paperless, unless you're sure you've authenticated that user.

The backend will create the submitted user if it doesn't exist already.

**Reply by skuzzle** [reply]
Ah Ok. Maybe I misinterpreted the Django documentation here. I'm not familiar with python.
Unfortunately I'm currently not able to try that out in my environment. The OAuth overlay container always passes the user as `X-Forwarded-User` and that header name can not be changed. If paperless offered a way to customize the header name that would be great. I might also create a ticket at https://github.com/thomseddon/traefik-forward-auth and request support for different header names.

**Reply by TheFehr** [reply]
@skuzzle you could also potentially use this: https://github.com/tomMoulard/htransformation
With this you could rename the X-Forwarded-User Header to Remote-User Header. It uses the new traefik plugin system.

**Reply by skuzzle** [reply]
That looks promising, thanks for the suggestion. I'm not particularly happy though to connect my personal traefik to another cloud service (_Traefik Pilot_ seems to be mandatory here) in order to use the plugin. But if no other solution comes up I will have to give it a shot any way.

**Reply by TheFehr** [reply]
Yeah that is the only reason so far keeping me from experimenting with some of the plugins that are out there. Just thought I'd throw that link here for anyone who stumbles upon this discussion.

**Reply by shamoon** [reply]
Traefik header transformation is definitely overdue, would solve so many of these (issue has been open since 2019 🧐 https://github.com/traefik/traefik/issues/6047 )

FWIW I use Authelia and implemented header auth for that and it wasn’t so bad, did you already look at #260 ?

**Reply by skuzzle** [reply]
The PR looks pretty simple and straight forward. Seems like allowing a configurable custom header name here is a no-brainer, at least for someone who knows python. I don't, but I'd be willing to take care of that feature unless one of you profis wants to jump in to safe the world :)

**Reply by skuzzle** [reply]
FYI: I've actually started a PR for this [here](https://github.com/jonaswinkler/paperless-ng/pull/677).

**Comment by dkowis** [comment]
There's also actual openid-connect, and using actual oauth2 with JWTs to talk to the API. This can also be done natively in angular using something like: https://www.npmjs.com/package/angular-auth-oidc-client where you authenticate directly against an openID connect service, and then the JWT is passed to the backend. 

I've done stuff like this professionally, but maybe it's more complicated than simple header things. I run a keycloak in my homelab, and so it'd be awesome to take advantage of openid-connect.

**Comment by telvgehc** [comment]
Update on this?

**Comment by scooper4711** [comment]
I too would like this feature. I'm running paperless on my Synology, which can be configured with SSO Server. I'd like to have paperless secured with 2FA (also with Synology) since I keep sensitive documents on it.

**Comment by mwllgr** [comment]
If you came from Google, paperless-ngx is working on OIDC support right now: https://github.com/paperless-ngx/paperless-ngx/pull/1746

**Reply by dkowis** [reply]
Today I learned about paperless-ngx! Thanks!
